experimental/oak_baremetal_simple_io/src/lib.rs

```diff
@@ -45,51 +47,77 @@ pub const OUTPUT_BUFFER_LEGNTH: usize = 4096;
 /// The length of the buffer that will be used for input messages.
 pub const INPUT_BUFFER_LEGNTH: usize = 4096;
 
+/// A Simple IO device implementation that uses direct port-based IO.
+pub type RawSimpleIo<'a> = SimpleIo<'a, PortReadOnly<u32>, PortWriteOnly<u32>>;
+
 /// The simple I/O channel driver implementation.
-pub struct SimpleIo {
+pub struct SimpleIo<'a, R: PortReader<u32> + 'a, W: PortWriter<u32> + 'a> {
     output_buffer: Vec<u8>,
     input_buffer: Vec<u8>,
-    output_length_port: PortWriteOnly<u32>,
-    input_length_port: PortReadOnly<u32>,
+    output_length_port: W,
+    input_length_port: R,
+    _phantom: PhantomData<&'a W>,
 }
```

Comment:
The lifetime bound looks a bit odd in the way it's used. Would something like this work?

```rust
pub struct SimpleIo<'a, R: PortReader<u32>, W: PortWriter<u32>> {
  ...
  output_length_port: &'a W,
  input_length_port: &'a R,
  ...
}
```

Also I wonder if that removes the need of the phantom data.

Reply:
The SimpleIo struct owns the two ports. If I change these to be references I get an error that the values are dropped while they are being borrowed, as nothing owns them.

The reason for the lifetime paramter is because the ports have references to the GHCB that in turn needs a lifetime parameter.

I don't like how this turned out with the complications around lifetime parameters etc. I am considering making the lifetime static and using `lazy_static` for the GHCB, seeing that the GHCB for a specific vCPU will be fixed for its entire life. But if I do that I will do it as a separate PR.